stixcore/products/level0/science.py

```diff
@@ -0,0 +1,275 @@
+                          ssid=self.ssid, data=data, ctronrol=control)
+
+    def __repr__(self):
+        return f'<{self.__class__.__name__}\n' \
+               f' {self.control.__repr__()}\n' \
+               f' {self.data.__repr__()}\n' \
+               f'>'
+
+    # @staticmethod
+    # def get_energies():
+    #     return get_energies_from_mask()
+    #
+    # @classmethod
+    # def from_fits(cls, fitspath):
+    #     header = fits.getheader(fitspath)
+    #     control = QTable.read(fitspath, hdu='CONTROL')
+    #     data = QTable.read(fitspath, hdu='DATA')
+    #     obs_beg = Time(header['DATE_OBS'])
+    #     data['time'] = (data['time'] + obs_beg)
+    #     return cls(control=control, data=data)
+
+    def to_requests(self):
+        for ci in unique(self.control, keys=['tc_packet_seq_control', 'request_id'])['index']:
+            control = self.control[self.control['index'] == ci]
+            data = self.data[self.data['control_index'] == ci]
+            # for req_id in self.control['request_id']:
+            #     ctrl_inds = np.where(self.control['request_id'] == req_id)
+            #     control = self.control[ctrl_inds]
+            #     data_index = control['index'][0]
+            #     data_inds = np.where(self.data['control_index'] == data_index)
+            #     data = self.data[data_inds]
+
+            yield type(self)(service_type=self.service_type, service_subtype=self.service_subtype,
+                             ssid=self.ssid, control=control, data=data)
+
+
+class XrayL0(ScienceProduct):
+    def __init__(self, *, service_type, service_subtype, ssid, control, data, **kwargs):
+        super().__init__(service_type=service_type, service_subtype=service_subtype,
+                         ssid=ssid, control=control, data=data, **kwargs)
+        self.name = 'xray-l1'
+        self.level = 'L1'
+
+    @classmethod
+    def from_levelb(cls, levelb):
+        packets = [Packet(unhexlify(d)) for d in levelb.data['data']]
+        packets = PacketSequence(packets)
+
+        service_type = packets.get('service_type')[0]
+        service_subtype = packets.get('service_subtype')[0]
+        ssid = packets.get('pi1_val')[0]
+
+        control = ControlSci.from_packets(packets)
+
+        # control.remove_column('num_structures')
+        control = unique(control)
+
+        if len(control) != 1:
+            raise ValueError('Creating a science product form packets from multiple products')
+
+        control['index'] = 0
+
+        structure = defaultdict(list)
+        subc_nixs = packets.get('NIX00403')[0][0].__dict__.keys()
+        for substructures in packets.get('NIX00403'):
+            for substructure in substructures:
+                for name in subc_nixs:
+                    val = getattr(substructure, name)
+                    try:
+                        structure[name].extend(val)
+                    except TypeError:
+                        structure[name].append(val)
```

Comment:
For rest of the code to work as is, which isn't necessity a great idea, this needs to be done. It probably super slow and I think could be done earlier in the packet flattening?